File: c/assembly.c
#include "assembly.h"
#include "bytecode.h"
#include "error.h"

typedef struct {
    int r;
    int g;
    int b;
    int a;
} register_t;

int imgasm_assemble_image(imgasm_template_t* template, int frame, imgasm_assembly_list_t* images) {
    if (!template || !images) {
        return imgasm_set_error(IMGASM_ERROR_INVALID_ASSEMBLY_LIST, "Invalid arguments");
    }
    
    if (frame < 0 || frame >= template->frame_count) {
        return imgasm_set_error(IMGASM_ERROR_INVALID_ASSEMBLY_LIST,
                                "Invalid frame index %d, index allowed between 0 and %d",
                                frame, template->frame_count);
    }
    
    if (images->image_count != template->input_count) {
        return imgasm_set_error(IMGASM_ERROR_INVALID_ASSEMBLY_LIST,
                                "Invalid image list length %d, needs to be %d",
                                images->image_count, template->input_count);
    }
    
    for (int i = 0; i < images->image_count; i++) {
        int pixels_in_frame = template->inputs[i].width * template->inputs[i].height;
        int pixels_in_input = (images->images[i].width * images->images[i].height);
        
        if (pixels_in_frame != pixels_in_input) {
            return imgasm_set_error(IMGASM_ERROR_INVALID_ASSEMBLY_IMAGE,
                                    "Invalid input index %d size %d, needs to be %d",
                                    i, pixels_in_input, pixels_in_frame);
        }
    }
    
    imgasm_frame_t* frm = &template->frames[frame];
    
    if (!frm->bytecode) {
        return imgasm_set_error(IMGASM_ERROR_INVALID_ASSEMBLY_IMAGE,
                                "Frame %d bytecode not loaded", frame);
    }
    
    unsigned char* output = images->images[0].pixels;
    int pixel_count = frm->width * frm->height;
    unsigned char* bytecode = (unsigned char*)frm->bytecode;
    int byte_index = 0;
    
    register_t registers[4];
    
    for (int pixel = 0; pixel < pixel_count; pixel++) {
        for (int i = 0; i < 4; i++) {
            registers[i].r = 0x00;
            registers[i].g = 0x00;
            registers[i].b = 0x00;
            registers[i].a = 0x00;
        }
        
        for (;;) {
            unsigned char operation = bytecode[byte_index];
            byte_index++;
            
            switch (operation) {
                case IMGASM_OP_NOOP:
                    break;
                    
                case IMGASM_OP_MOVE: {
                    unsigned char param = bytecode[byte_index];
                    byte_index++;
                    
                    unsigned char destination_register = param >> 4;
                    unsigned char source_register = param & 0x0F;
                    
                    registers[destination_register].r = registers[source_register].r;
                    registers[destination_register].g = registers[source_register].g;
                    registers[destination_register].b = registers[source_register].b;
                    registers[destination_register].a = registers[source_register].a;
                    break;
                }
                
                case IMGASM_OP_RETURN:
                    break;
                    
                case IMGASM_OP_CONSTANT: {
                    unsigned char param = bytecode[byte_index];
                    byte_index++;
                    
                    unsigned char destination_register = param >> 4;
                    
                    registers[destination_register].r = bytecode[byte_index + 0];
                    registers[destination_register].g = bytecode[byte_index + 1];
                    registers[destination_register].b = bytecode[byte_index + 2];
                    registers[destination_register].a = bytecode[byte_index + 3];
                    
                    byte_index += 4;
                    break;
                }
                
                case IMGASM_OP_SAMPLE: {
                    unsigned char param = bytecode[byte_index];
                    byte_index++;
                    
                    unsigned char destination_register = param >> 4;
                    unsigned char source_input = param & 0x0F;
                    
                    unsigned short x = bytecode[byte_index + 0] | (bytecode[byte_index + 1] << 8);
                    unsigned short y = bytecode[byte_index + 2] | (bytecode[byte_index + 3] << 8);
                    
                    byte_index += 4;
                    
                    imgasm_image_t* source_image = &images->images[source_input];
                    int source_pixel = source_image->width * y + x;
                    int source_index = source_pixel * 4;
                    
                    registers[destination_register].r = source_image->pixels[source_index + 0];
                    registers[destination_register].g = source_image->pixels[source_index + 1];
                    registers[destination_register].b = source_image->pixels[source_index + 2];
                    registers[destination_register].a = 0xFF;
                    break;
                }
                
                case IMGASM_OP_ADD: {
                    unsigned char param = bytecode[byte_index];
                    byte_index++;
                    
                    unsigned char destination_register = param >> 4;
                    unsigned char source_register = param & 0x0F;
                    
                    registers[destination_register].r += registers[source_register].r;
                    registers[destination_register].g += registers[source_register].g;
                    registers[destination_register].b += registers[source_register].b;
                    break;
                }
                
                case IMGASM_OP_MULTIPLY: {
                    unsigned char param = bytecode[byte_index];
                    byte_index++;
                    
                    unsigned char destination_register = param >> 4;
                    unsigned char source_register = param & 0x0F;
                    
                    registers[destination_register].r *= registers[source_register].r;
                    registers[destination_register].g *= registers[source_register].g;
                    registers[destination_register].b *= registers[source_register].b;
                    
                    registers[destination_register].r >>= 8;
                    registers[destination_register].g >>= 8;
                    registers[destination_register].b >>= 8;
                    break;
                }
                
                default:
                    break;
            }
            
            if (operation == IMGASM_OP_RETURN) {
                break;
            }
        }
        
        if (registers[0].r > 255) registers[0].r = 255;
        if (registers[0].g > 255) registers[0].g = 255;
        if (registers[0].b > 255) registers[0].b = 255;
        if (registers[0].a > 255) registers[0].a = 255;
        
        output[pixel * 4 + 0] = (unsigned char)registers[0].r;
        output[pixel * 4 + 1] = (unsigned char)registers[0].g;
        output[pixel * 4 + 2] = (unsigned char)registers[0].b;
        output[pixel * 4 + 3] = 0xFF;
    }
    
    imgasm_clear_error();
    return IMGASM_ERROR_NONE;
}

File: dart/lib/assembly.dart
import 'dart:typed_data';
import 'template.dart';
import 'bytecode.dart';

class ImgasmImage
{
    int width;
    int height;
    Uint8List pixels;

    ImgasmImage(this.width, this.height, this.pixels);
}

class Interpreter 
{
  static ImgasmImage assemble(Template template, int frameIndex, List<ImgasmImage> inputs) 
  {
    if (frameIndex < 0 || frameIndex >= template.frames.length) 
    {
      throw Exception("Invalid frame index: $frameIndex");
    }

    if (inputs.length != template.inputs.length) 
    {
      throw Exception("Invalid input count: expected ${template.inputs.length}, got ${inputs.length}");
    }

    Frame frame = template.frames[frameIndex];

    Uint8List outputPixels = Uint8List(frame.width * frame.height * 4);

    final List<Int32List> registers = List.generate(16, (_) => Int32List(4));


    int byteIndex = 0; 
    Uint8List bc = frame.bytecode!;
    int totalPixels = frame.width * frame.height;

    for (int p = 0; p < totalPixels; p++) 
    {
        for (var reg in registers) 
        {
            reg[0] = reg[1] = reg[2] = reg[3] = 0;
        }

        while (byteIndex < bc.length)
        {
            int operation = bc[byteIndex++];

            if (operation == OpCode.ret) break; 
            switch (operation) 
            {
            case OpCode.noop:
                break;

            case OpCode.move:
                int param = bc[byteIndex++];
                int dst = param >> 4;
                int src = param & 0x0F;
                registers[dst][0] = registers[src][0];
                registers[dst][1] = registers[src][1];
                registers[dst][2] = registers[src][2];
                registers[dst][3] = registers[src][3];
                break;

            case OpCode.constant:
                int param = bc[byteIndex++];
                int dst = param >> 4;
                registers[dst][0] = bc[byteIndex++];
                registers[dst][1] = bc[byteIndex++];
                registers[dst][2] = bc[byteIndex++];
                registers[dst][3] = bc[byteIndex++];
                break;

            case OpCode.sample:
                int param = bc[byteIndex++];
                int dst = param >> 4;
                int inputIdx = param & 0x0F;
                int sx = bc[byteIndex++] | (bc[byteIndex++] << 8);
                int sy = bc[byteIndex++] | (bc[byteIndex++] << 8);

                if (inputIdx < inputs.length) {
                ImgasmImage srcImg = inputs[inputIdx];
                if (sx >= 0 && sx < srcImg.width && sy >= 0 && sy < srcImg.height) {
                    int srcIdx = (sy * srcImg.width + sx) * 4;
                    registers[dst][0] = srcImg.pixels[srcIdx + 0];
                    registers[dst][1] = srcImg.pixels[srcIdx + 1];
                    registers[dst][2] = srcImg.pixels[srcIdx + 2];
                    registers[dst][3] = 255;
                }
                }
                break;

            case OpCode.add:
                int param = bc[byteIndex++];
                int dst = param >> 4;
                int src = param & 0x0F;
                registers[dst][0] += registers[src][0];
                registers[dst][1] += registers[src][1];
                registers[dst][2] += registers[src][2];
                break;

            case OpCode.multiply:
                int param = bc[byteIndex++];
                int dst = param >> 4;
                int src = param & 0x0F;
                registers[dst][0] = (registers[dst][0] * registers[src][0]) >> 8;
                registers[dst][1] = (registers[dst][1] * registers[src][1]) >> 8;
                registers[dst][2] = (registers[dst][2] * registers[src][2]) >> 8;

                break;
                
            default:
                break;
            }
        }

        int outIdx = p * 4;
        outputPixels[outIdx + 0] = registers[0][0].toUnsigned(8);
        outputPixels[outIdx + 1] = registers[0][1].toUnsigned(8);
        outputPixels[outIdx + 2] = registers[0][2].toUnsigned(8);
        outputPixels[outIdx + 3] = 255; 
        }


    return ImgasmImage(frame.width, frame.height, outputPixels);
  }
}